### ISSUE TITLE
Linq-to-XML difficult to use with newly added nullability annotations

### ISSUE THREAD

**Issue by chrarnoldus** [dotnet/runtime#42316]
### Description

RC1 added nullability annotations for System.Xml.Linq, which make extensive use of `[NotNullIfNotNull(...)]` on user defined conversions. For example:

https://github.com/dotnet/runtime/blob/33dba9518b4eb7fbc487fadc9718c408f95a826c/src/libraries/System.Private.Xml.Linq/src/System/Xml/Linq/XName.cs#L106

This is not yet supported by the compiler, however (see https://github.com/dotnet/roslyn/issues/32671), resulting in false positives being reported. Some examples are included below:

```c#
#nullable enable
using System;
using System.Xml.Linq;

namespace ConsoleApp2
{
    internal class Program
    {
        private static void Main()
        {
            // Reports "warning CS8604: Possible null reference argument for parameter 'name' in 'XElement.XElement(XName name, object? content)'."
            // for "a", which obviously isn't null.
            // Can be worked around using XName.Get.
            var el = new XElement("a", new XAttribute(XName.Get("b"), "c"));

            // Reports "warning CS8600: Converting null literal or possible null value to non-nullable type.",
            // despite throwing an exception when the XAttribute is null.
            var c = (string) (el.Attribute(XName.Get("b")) ?? throw new Exception());

            Console.WriteLine(c);
        }
    }
}
```

I hope something can be done to mitigate this situation, because having to change code that has already been updated to use nullable reference types in order to work around a compiler limitation would be unfortunate.

### Configuration

```
.NET SDK (reflecting any global.json):
 Version:   5.0.100-rc.1.20452.10
 Commit:    473d1b592e

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19041
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\5.0.100-rc.1.20452.10\

Host (useful for support):
  Version: 5.0.0-rc.1.20451.14
  Commit:  38017c3935
```

### Regression?

Yes, compared to preview8.

**Comment by stephentoub** [comment]
cc: @jcouv

**Comment by jeffhandley** [comment]
What do you suggest we do on this, @krwq, @buyaa-n, and @stephentoub?

**Comment by stephentoub** [comment]
My understanding is the annotation is correct but the compiler isn't yet doing the right thing. If that's true, the right fix is in the compiler, not to degrade the APIs. @jcouv, @cston, is my understanding correct, and if so, when will the compiler fix this?

**Comment by jeffhandley** [comment]
I'm applying the blocking-release label on this one so that we can track it closely and make a deliberate decision before 5.0 ships.

**Comment by stephentoub** [comment]
@chsienki, @Jaredpar, could you comment on the status of https://github.com/dotnet/roslyn/issues/32671?

**Comment by RikkiGibson** [comment]
The scenario in this issue's description (`[NotNullIfNotNull]` in user-defined conversion) is fixed by dotnet/roslyn#47555.

However the exact scenario in dotnet/roslyn#32671 is not fixed--I forgot to test the scenario of a conversion to bool with attributes like `[NotNullWhen]` on the parameter 😅

**Comment by stephentoub** [comment]
Thanks, @RikkiGibson.  Sounds like this can be closed then, and any straggling issues will be handled by the compiler. Thanks!
